### PR TITLE
feat(optimizer): Add output indices to `LogicalJoin`

### PR DESCRIPTION
**Pull request by jon-chuang**
## What's changed and what's your intention?
Adds output indices to `LogicalJoin` to avoid simple projections which merely reorder columns.

Future:
- Add more optimizer rules to merge simple projections into `LogicalJoin`s/other operators with `output_indices`.
- Impl output_indices in plan proto and executors. For streaming hash join, we can just modify the chunk builder to take in an `output_indices` parameter.
- `LogicalMultiJoin` also has `output_indices`. It needs to merge output indices and rewrite conditions. 

## Refer to a related PR or issue link (optional)
Tracked in: https://github.com/singularity-data/risingwave/issues/1922

### CODE REVIEW

**Review by Enter-tainer — APPROVED** [review]
Generally LGTM.

BTW, just to confirm, for o2l l2o r2o o2r... mappings, the "o" here should be "original" rather than "output", right? Maybe the comments should be tweaked. 

cc @st1page PTAL.

**Comment by jon-chuang** [comment]
> BTW, just to confirm, for o2l l2o r2o o2r... mappings, the "o" here should be "original" rather than "output", right? Maybe the comments should be tweaked.

Yes, I think I try to maintain the concept of `internal_column_number`. Perhaps it should be renamed to `i2l`, `l2i`, and we also introduce `i2o` which is the mapping defined by the `output_indices`. I used this concept of "internal" columns in certain parts of this PR. I think I'll adopt it and add some comments.

File: src/frontend/src/optimizer/plan_node/logical_join.rs
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:786]: Add comments that we are adding PK indices to the output_indices

File: src/frontend/src/optimizer/plan_node/logical_join.rs
Comment (Enter-tainer) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:69]: `output_indices` can not only handle reorder but also column removals. Maybe `new_with_output_indices`?

**Comment by st1page** [comment]
>Yes, I think I try to maintain the concept of internal_column_number. Perhaps it should be renamed to i2l, l2i, and we also introduce i2o which is the mapping defined by the output_indices. I used this concept of "internal" columns in certain parts of this PR. I think I'll adopt it and add some comments.

could you please do it in this PR? I am afraid that some others use the "o2l, l2o" such as distribution derive or some rules. so we might be opposed to guaranteed current semantics.

**Comment by jon-chuang** [comment]
Hi! ready for final review. I applied suggestions from the comments and also added some changes for affected components.

File: src/frontend/src/optimizer/plan_node/logical_join.rs
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:663]: Nice work! I did not think about that.

**Review by st1page — APPROVED** [review]
LGTM

File: src/frontend/test_runner/tests/testdata/join.yaml
Comment (skyzh) [inline src/frontend/test_runner/tests/testdata/join.yaml:11]: I think this should only be shown when it isn't `0, 1, 2, ...`

**Comment by skyzh** [comment]
... and there's some problems with the plan in TPC-H:

```
                        StreamProject { exprs: [$1, $0, $2, $3, $4, $5] }
                           StreamProject { exprs: [$0, $1, $3, $4, $5, $7] }
```

It seems that two projections are not combined? Is it because of rule application order?

**Comment by Enter-tainer** [comment]
> It seems that two projections are not combined? Is it because of rule application order?

A workaround is discussed in https://github.com/singularity-data/risingwave/pull/2769#discussion_r883316583 . And I'm working on pushing output indices to executors so the workaround will not last too long.

File: src/frontend/test_runner/tests/testdata/join.yaml
Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/join.yaml:11]: makes sense!

**Comment by jon-chuang** [comment]
Yes, basically its due to a workaround!

**Comment by skyzh** [comment]
So we should merge https://github.com/singularity-data/risingwave/pull/2769 before merging this PR? 🤪

**Comment by Enter-tainer** [comment]
I think the order of merging is not a big issue because extra batch/stream project nodes don't hurt correctness :rofl: . And I agree with you that it would be better if we merge #2769 before this pr. BTW, works in #2769 are mostly done and hopefully can be merged soon. But if this PR is ready, I think it is OK to merge it.

**Comment by jon-chuang** [comment]
should be ready, I'll merge when the CI passes!

**Pull request merged**
